File: 백준/Silver/17086. 아기 상어 2/아기 상어 2.js
const input = require("fs")
  .readFileSync("/dev/stdin", "utf8")
  .trim()
  .split("\n")
  .map((line) => line.replace(/\r/, ""));

// 첫번째줄에 rown, colN이 주어지고 나머지에 그리드가 주어진다
// 0은 빈칸 1은 아기상어. 상하좌우 대각선으로 이동이 가능하고
// 안전거리가 가장 큰 거리를 구해라
// 먼지 dist로 모든 값을 Infinty로 초기한 후 아기상어 좌표는 0으로 둔다
// 다음 거리가 d+1 가 작으면 이동한다
const solution = (input) => {
  const [rowN, colN] = input[0].split(" ").map((v) => Number(v));
  const grid = [];
  for (let i = 1; i <= rowN; i++) {
    grid.push(input[i].split(" ").map((v) => Number(v)));
  }

  const distGrid = Array(rowN)
    .fill(0)
    .map((v) => Array(colN).fill(Infinity));

  const queue = [];

  for (let i = 0; i < rowN; i++) {
    for (let j = 0; j < colN; j++) {
      if (grid[i][j] === 1) {
        queue.push([i, j, 0]);
        distGrid[i][j] = 0;
      }
    }
  }

  const dirs = [
    [-1, 0],
    [1, 0],
    [0, -1],
    [0, 1],
    [1, -1],
    [1, 1],
    [-1, -1],
    [-1, 1],
  ];

  let answer = 0;
  while (queue.length) {
    const [x, y, d] = queue.shift();
    answer = Math.max(answer, d);

    for (const [dx, dy] of dirs) {
      const nextX = x + dx;
      const nextY = y + dy;
      const isValid =
        nextX >= 0 &&
        nextX < rowN &&
        nextY >= 0 &&
        nextY < colN &&
        d + 1 < distGrid[nextX][nextY];

      if (isValid) {
        queue.push([nextX, nextY, d + 1]);
        distGrid[nextX][nextY] = d + 1;
      }
    }
  }

  console.log(answer);
};

solution(input);

File: 백준/Silver/17086. 아기 상어 2/아기 상어 2.py
from collections import deque

# 첫째줄에 rowN, colN이 주어진다
# 두번째줄부터 그리드가 0,1로 주어지고 1이 상어다
# 각 지점에 상하좌우 대각선까지 이동가능하고
# 안전거리가 가자으 큰 값을 구해야된다
# bfs로 탐색을 하면되는데 먼저 이중for문으로
# 1인 지점을 찾아서 큐에 넣고 visited로 마킹하면된다

rowN, colN = list(map(int, input().split()))
grid = []
for i in range(rowN):
  grid.append(list(map(int, input().split())))

queue = deque()
visited = [[False]*colN for _ in range(rowN)]

for i in range(rowN):
  for j in range(colN):
    if grid[i][j] == 1:
      queue.append([i,j,0])
      visited[i][j] = True

answer = 0
dirs = [[-1,-1],[-1,0],[-1,1],[0,-1],[0,1],[1,-1],[1,0],[1,1]]

while len(queue):
  x, y, dist = queue.popleft()
  answer = max(answer, dist)

  for dx, dy in dirs:
    nextX = x + dx
    nextY = y + dy
    isValid = 0<=nextX<rowN and 0<=nextY<colN and not visited[nextX][nextY]

    if isValid:
      queue.append([nextX, nextY, dist+1])
      visited[nextX][nextY] = True

print(answer)
